File: lib/reedit-versions.js
'use strict';

// RE-EDIT VERSIONS — the ordinal a user sees on a re-edited video, and the
// one-at-a-time guard that keeps two of them from claiming the same number.
//
// WHY THE ORDINAL IS NOT DEPTH IN THE CHAIN. Measured against prod 2026-09-21:
// 114 re-edit rows, 97 distinct parents, 29 whose parent is ITSELF a re-edit,
// and a maximum of 4 children hanging off a single parent. parent_job_id
// describes a TREE, not a line. "version = how many links back to the root"
// therefore hands two siblings the same number, and there are already parents
// with four of them. The ordinal has to be anchored at the ROOT.
//
// WHY THE ORDINAL IS NOT PERSISTED. The obvious move is an integer column
// written at insert. It cannot be both stable and gapless:
//   - number every row, and a failed re-edit burns a number: the list reads
//     v1, v2, v4 and the user asks what happened to v3;
//   - number only the ones that succeed, and a failure RENUMBERS everything
//     after it — a version the user opened as v3 yesterday is v2 today.
// Computing it on read over COMPLETED rows only gives both at once, and it is
// safe precisely because terminal states are terminal here (first-terminal-wins
// on the way in, dispatch-to-modal.js): a completed job never un-completes, so
// a number, once shown, never moves.
//
// A FAILED RE-EDIT IS NOT A VERSION. It produced nothing the user can open.
// It belongs in the chat as a failed attempt, never in the version switcher —
// which is also why `version_count` counts completed rows and nothing else.

// NAMED FOR THE PROPERTY, NOT THE SCHEDULE. The first version of this called
// itself IN_FLIGHT_STATUSES and meant "non-terminal". That was correct until it
// wasn't: `needs_input` is non-terminal with NOTHING RUNNING, and the moment the
// client learned to answer a parked question (1e7aa37) the two stopped being the
// same fact. A guard whose membership is right when written and is invalidated by
// a change somewhere else is exactly what a schedule-shaped name hides — nothing
// in "in flight" announced which property a caller was relying on.
//
// So the two reasons a row holds its root are now separate and each says why.

// A worker is running, or is about to be. Matches the set the rest of the
// codebase already uses for this (job-reaper.js, orphan-redispatch.js).
const RENDERING_STATUSES = Object.freeze(['queued', 'processing']);

// Waiting on the USER, with no worker running and no slot held. Two different
// mechanisms land here and only one of them is live (see lib/clarification.js):
// the dormant Phase D ask-back, and the plan-diff clarification that has parked
// 19 rows across 8 users for up to 63 days.
const PARKED_STATUSES = Object.freeze(['needs_input']);

// What forbids a SECOND re-edit on the same root — Zac's "one re-edit at a time
// per video". A parked question is included deliberately: if it were not, a user
// sitting on an unanswered question could start a second render of the same
// video, which is the case the rule exists to prevent.
//
// The escape is not expiry. Answering the question CANCELS the parked row in the
// same operation that creates the retry (cancelAndCreate below), so a user who
// answers never meets this lock at all; the 24h expiry only catches the user who
// walked away. That ordering is load-bearing — see the note on provisionalVersion.
const ROOT_LOCK_STATUSES = Object.freeze([...RENDERING_STATUSES, ...PARKED_STATUSES]);

function isRendering(status) {
  return RENDERING_STATUSES.includes(String(status || ''));
}

function isParked(status) {
  return PARKED_STATUSES.includes(String(status || ''));
}

function holdsRootLock(status) {
  return ROOT_LOCK_STATUSES.includes(String(status || ''));
}

// Retained as an alias so an existing caller cannot silently change meaning on
// this rename. New code should say which property it means.
const IN_FLIGHT_STATUSES = ROOT_LOCK_STATUSES;
const isInFlight = holdsRootLock;

// Deterministic order. created_at ascending, id as the tie-break — without the
// tie-break, two rows written in the same millisecond (which the branched rows
// in prod already demonstrate is reachable) could order differently between two
// reads and silently renumber the list.
function byCreatedThenId(a, b) {
  const ta = Date.parse(a && a.created_at) || 0;
  const tb = Date.parse(b && b.created_at) || 0;
  if (ta !== tb) return ta - tb;
  return String((a && a.id) || '').localeCompare(String((b && b.id) || ''));
}

/**
 * The rows that ARE versions: completed, in display order.
 * @param {Array<object>} rows every row sharing one root_job_id
 */
function completedVersions(rows) {
  return (Array.isArray(rows) ? rows : [])
    .filter((r) => r && r.status === 'completed')
    .sort(byCreatedThenId);
}

/**
 * job_id -> 1-based dense version number. Completed rows only; a row that is
 * not a version is absent from the map rather than mapped to null, so a caller
 * that forgets to check gets `undefined` and not a plausible-looking 0.
 */
function versionMap(rows) {
  const map = new Map();
  completedVersions(rows).forEach((r, i) => map.set(String(r.id), i + 1));
  return map;
}

function versionOf(rows, jobId) {
  const v = versionMap(rows).get(String(jobId));
  return v === undefined ? null : v;
}

function versionCount(rows) {
  return completedVersions(rows).length;
}

/**
 * The in-flight row for this root, or null. There is at most one once the
 * partial unique index is live; before then this is still the read the 409
 * answers from, and it takes the OLDEST so the answer is stable across polls
 * rather than flapping between two rows.
 */
function findInFlight(rows) {
  const live = (Array.isArray(rows) ? rows : [])
    .filter((r) => r && holdsRootLock(r.status))
    .sort(byCreatedThenId);
  return live.length ? live[0] : null;
}

/**
 * The parked rows on this root that a reply should clear, oldest first.
 *
 * Only PARKED rows — a rendering job is never cancelled by someone answering a
 * question, because something is actually running and cancelling it would throw
 * away work the user is waiting for.
 */
function findParked(rows) {
  return (Array.isArray(rows) ? rows : [])
    .filter((r) => r && isParked(r.status))
    .sort(byCreatedThenId);
}

/**
 * The version a job that is still rendering WILL occupy if it completes.
 *
 * Exact rather than a guess, and only because of the 409: one re-edit at a time
 * per root means nothing else can complete and take the number first. If the
 * one-at-a-time rule is ever relaxed, this becomes a race and the caller must
 * stop flagging it `version_provisional` and start returning null.
 */
function provisionalVersion(rows) {
  return versionCount(rows) + 1;
}

/**
 * The versions[] payload, minus the signed urls the caller mints on read.
 * Completed rows only — see the header note on why a failure is not a version.
 */
function buildVersionList(rows) {
  return completedVersions(rows).map((r, i) => ({
    job_id: r.id,
    version: i + 1,
    status: r.status,
    created_at: r.created_at,
    change_request: r.change_request || null,
  }));
}

/**
 * Root for a row. An original IS its own root; a re-edit inherits its parent's.
 *
 * Returns null when the row is a re-edit whose root_job_id has not been
 * backfilled — the caller must walk parent_job_id rather than invent a root.
 * Guessing `id` there would make a v3 its own root and strand v1 and v2.
 */
function resolveRootId(row) {
  if (!row) return null;
  if (row.root_job_id) return String(row.root_job_id);
  if (!row.parent_job_id) return String(row.id);
  return null;
}

module.exports = {
  RENDERING_STATUSES,
  PARKED_STATUSES,
  ROOT_LOCK_STATUSES,
  isRendering,
  isParked,
  holdsRootLock,
  findParked,
  IN_FLIGHT_STATUSES,
  isInFlight,
  byCreatedThenId,
  completedVersions,
  versionMap,
  versionOf,
  versionCount,
  findInFlight,
  provisionalVersion,
  buildVersionList,
  resolveRootId,
};

File: ios/Promptly/Promptly/Models/ReeditVersions.swift
import Foundation

// RE-EDIT VERSIONS — the wire shape, and the invariants that come with it.
//
// A video's history is a FLAT LIST under a root, not a tree. Re-edits branch in
// the data (parent_job_id records provenance, and 114 real re-edits already
// include chains four deep and parents with four children), but the user sees
// creation order under the root: v1, v2, v3, latest last. That is why the
// ordinal is root-anchored and NOT depth — two siblings at the same depth would
// otherwise both be "version 2", and there are already parents with four of them.
//
// ORDINALS ARE COMPUTED, NOT STORED, and they count COMPLETED rows only. A
// persisted integer cannot be both stable and gapless: number every attempt and
// a failure burns v3 forever; number only successes and a failure renumbers what
// the user opened yesterday. Computed over completed rows, ordered created_at
// with id as tie-break, it is both — and the tie-break is load-bearing, because
// two rows have already been written in the same millisecond in production.
//
// A FAILED JOB IS NOT A VERSION. It produced nothing to open. It belongs in the
// job's chat as a failed attempt and never in the switcher, which is why
// `version` is null for it and why version_count counts completions.




/// The 409 from POST /api/video-jobs/re-edit. Carries `status` so the composer
/// reads its state rather than inferring it: parked-on-a-question and
/// actively-rendering look identical from a bare 409 and need different copy.
struct ReeditInFlight: Decodable, Equatable {
    let error: String
    let status: String              // "queued" | "processing" | "needs_input"
    /// The live job holding the root. The server spells this `job_id`; this
    /// model spelled it `in_flight_job_id` and the two never met.
    let jobId: String?
    let rootJobId: String?
    /// The ordinal the live job will take if it completes.
    let version: Int?

    /// EVERY FIELD BUT `status` IS OPTIONAL, ON PURPOSE.
    ///
    /// This decoded behind `try?` in APIService — so a shape mismatch did not
    /// throw, it produced nil, the `if let` fell through, and the caller got a
    /// generic "Re-edit failed". The typed 409 that the composer's parked and
    /// rendering states are built on would simply never have been produced, and
    /// nothing would have said so. Verified against the real body before fixing:
    ///   keyNotFound "in_flight_job_id" — the server sends `job_id`.
    ///
    /// `status` is the only field that changes what the user is told, so losing
    /// an informational id must never cost the typed error again.
    init(from decoder: Decoder) throws {
        let c = try? decoder.container(keyedBy: CodingKeys.self)
        error  = ((try? c?.decodeIfPresent(String.self, forKey: .error)) ?? nil) ?? "reedit_in_flight"
        status = ((try? c?.decodeIfPresent(String.self, forKey: .status)) ?? nil) ?? ""
        // Accept the server's spelling first, then this model's old one, so a
        // client and a server that disagree still produce a typed error.
        let id = ((try? c?.decodeIfPresent(String.self, forKey: .job_id)) ?? nil)
            ?? ((try? c?.decodeIfPresent(String.self, forKey: .in_flight_job_id)) ?? nil)
        jobId = id
        rootJobId = ((try? c?.decodeIfPresent(String.self, forKey: .root_job_id)) ?? nil)
        version = ((try? c?.decodeIfPresent(Int.self, forKey: .version)) ?? nil)
    }

    enum CodingKeys: String, CodingKey {
        case error, status, job_id, in_flight_job_id, root_job_id, version
    }

    /// needs_input is IN the in-flight set deliberately — 19 rows in production,
    /// every one a re-edit, 17% of all re-edits. A user sitting on an unanswered
    /// question could otherwise start a second render of the same video. The
    /// escape is answering the question or cancelling it, not a second re-edit.
    var isParkedOnAQuestion: Bool { status == "needs_input" }
    var isRendering: Bool { status == "queued" || status == "processing" }
}


/// AN OFFER OF REFINEMENT, POSTED AFTER THE VIDEO IS DELIVERED.
///
/// NOT a ParkedClarification, and the difference is the whole design. A parked
/// clarification is a job STOPPED at `needs_input`, waiting to be unblocked —
/// the question gates the work. This is its opposite: B1's capture shows the
/// agent asking at the END of its turn (blocks 16/26 and 24/38), after the work
/// is done, as an offer — "Captions are now live… Would you like any emphasis
/// added to specific words?" The video is already delivered and keeps its
/// delivered state whatever the user does.
///
/// Because they are opposites, they cannot share a render path: the existing
/// ClarificationCard draws only when `jobStatus == "needs_input"`, which a
/// delivered job never is. Reusing it would have shown nothing at all.
struct RefinementOffer: Codable, Equatable, Hashable {
    /// The question, verbatim from the agent. Rendered as ordinary assistant
    /// prose, not as a card with a Send button — it is a message, not a form.
    let question: String
    /// The job row this is about. PAIRED BY ROW, never by text or position, so
    /// a superseding re-edit can retract exactly this one.
    let jobId: String
    /// The version delivered when the question was asked. A question about v2
    /// must never appear under v3.
    let versionId: String?

    enum CodingKeys: String, CodingKey {
        case question, jobId = "job_id", versionId = "version_id"
    }

    init(question: String, jobId: String, versionId: String? = nil) {
        self.question = question
        self.jobId = jobId
        self.versionId = versionId
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        question = try c.decode(String.self, forKey: .question)
        jobId = try c.decode(String.self, forKey: .jobId)
        versionId = (try? c.decodeIfPresent(String.self, forKey: .versionId)) ?? nil
    }
}
